File: build_dataset_runXGB.R
library(plyr)
library(dplyr)
library(zoo)
library(MASS)
library(glmnet)
library(pROC)
library(caret)
library(ROCR)
library(xgboost)


`%notin%` <- Negate(`%in%`)
# import dataset
df = read.csv('COVID-19_Historical_Data_by_State.orig.csv')
df = df[order(df$DATE),]
row.names(df) <- NULL
# add new column HOSP_NEW as NA
df$HOSP_NEW = NA
# add HOSP_NEW values as difference between current day and prev day
df = df %>% mutate(HOSP_NEW = HOSP_YES - lag(HOSP_YES, default = HOSP_YES[1]))

# repeat for the values: HOSP_NO, HOSP_UNK, POS_FEM, POS_MALE
# then, create the label by scanning the next day to see if there is an increase in POS_NEW
df$SCAN = NA
df = df %>% mutate(SCAN = POS_NEW - lead(POS_NEW, default = POS_NEW))
View(df[which(colnames(df) %in% c('HOSP_NEW', 'HOSP_YES', 'SCAN', 'POS_NEW', 'INCREASE'))])
df$INCREASE = ifelse(df$SCAN < 0, 1.0, 0.0)
# select columns and rows, then you're ready to build the XGBoost model
df_colFiltered = df[which(names(df) %in% c("POS_NEW", "POS_7DAYAVG", "NEG_NEW", "NEG_7DAYAVG", "DTH_NEW", "DTH_7DAYAVG", "TEST_NEW", "TEST_7DAYAVG", "HOSP_NEW", "HOSP_NO_NEW", "HOSP_UNK_NEW", "POS_FEM_NEW", "POS_MALE_NEW", "INCREASE"))]
df_colFiltered = tail(df_colFiltered, -20)
df_colFiltered = head(df_colFiltered, -1)
row.names(df_colFiltered) <- NULL

df = df_colFiltered
df$INCREASE = ifelse(df$INCREASE == 1, "increasing", "decreasing")


df_split = sample.int(n=nrow(df), size=floor(.75*nrow(df)), replace=FALSE)
df_training = df[df_split,]
df_testing = df[-df_split,]


cvCtrl <- trainControl(method = "repeatedcv", number = 5, repeats = 1, classProbs = TRUE, summaryFunction = twoClassSummary, allowParallel = TRUE)

xgbGrid<-expand.grid(nrounds=c(500, 750), max_depth=c(2, 5), eta=c(0.001, 0.01), gamma = c(0), colsample_bytree = c(1), min_child_weight = c(2, 4), subsample = c(0.5))
xgbGridtrain<- train(INCREASE~., method = "xgbTree", data = df_training, metric = "ROC", tree_method = "auto",trControl = cvCtrl, tuneGrid = xgbGrid, na.action = na.pass)
xgbGridProbs<-predict(xgbGridtrain, newdata=df_testing, type="prob", na.action = na.pass)
xgbGridProbsDF<-as.data.frame(xgbGridProbs)

xgbGridRoc <-roc(df_testing$INCREASE, round(xgbGridProbsDF$increasing, digits=4))

print('baseline AUC is')
print(xgbGridRoc)

File: build_dataset_runXGB.py
import pandas as pd
import numpy as np
import xgboost
import random
from sklearn.metrics import accuracy_score

df = pd.read_csv('COVID-19_Historical_Data_by_State.orig.csv')

df = df.sort_values('DATE')
df = df.reset_index(drop=True)

# repeat for the values "POS_FEM_NEW", "POS_MALE_NEW", "HOSP_UNK_NEW",
df['HOSP_NEW'] = df['HOSP_YES'].diff()

# setup labels
df['SCAN'] = df['POS_NEW'].diff(periods=-1)
df['INCREASE'] = -1
df.loc[df['SCAN'] < 0, ['INCREASE']] = 1
df.loc[df['SCAN'] >= 0, ['INCREASE']] = 0

df = df.iloc[20:]
# df_filtered = df[["POS_NEW", "POS_7DAYAVG", "NEG_NEW", "NEG_7DAYAVG", "DTH_NEW", "DTH_7DAYAVG", "TEST_NEW", "TEST_7DAYAVG", "HOSP_NEW", "HOSP_NO_NEW", "HOSP_UNK_NEW", "POS_FEM_NEW", "POS_MALE_NEW", "INCREASE"]]
df_filtered.drop(df.tail(1).index, inplace=True)
df_filtered = df[["POS_NEW", "POS_7DAYAVG", "NEG_NEW", "NEG_7DAYAVG", "DTH_NEW", "DTH_7DAYAVG", "TEST_NEW", "TEST_7DAYAVG", "HOSP_NEW", "INCREASE"]]
df_filtered



df_split = np.random.rand(len(df_filtered)) < 0.8))
df_training = df_filtered[df_split]
df_testing = df_filtered[~df_split]
df_training_y = df_training['INCREASE']
df_training_x = df_training
del df_training_x['INCREASE']
df_training_y = df_training_y.tolist()

df_testing_y = df_testing['INCREASE']
df_testing_x = df_testing
del df_testing_x['INCREASE']
df_testing_y = df_testing_y.tolist()

model = xgboost.XGBClassifier()
model.fit(df_training_x, df_training_y)

y_pred = model.predict(df_testing_x)


predictions = [round(value) for value in y_pred]
accuracy = accuracy_score(df_testing_y, predictions)
print("Accuracy is %.2f%%" % (accuracy * 100.0))
